src/gf.c

```diff
@@ -1730,6 +1730,78 @@ DLLEXPORT jl_value_t *jl_gf_invoke_lookup(jl_function_t *gf, jl_datatype_t *type
     return (jl_value_t*)m;
 }
 
+jl_function_t*
+jl_gf_invoke_get_specialization(jl_function_t *gf, jl_tupletype_t *types,
```

Comment:
there seems to be a lot of code duplication between this method and `jl_gf_invoke`. can you refactor the two so that the common bits are shared?